src/client/microsoftAuth.js

```diff
@@ -1,26 +1,107 @@
 const XboxLiveAuth = require('@xboxreplay/xboxlive-auth')
+const msal = require('@azure/msal-node')
 const fetch = require('node-fetch')
+const { MsaTokenManager, XboxTokenManager, MinecraftTokenManager } = require('./tokens')
+
+const getFetchOptions = {
+  headers: {
+    'Content-Type': 'application/json',
+    'User-Agent': 'node-minecraft-protocol'
+  }
+}
 
 const XSTSRelyingParty = 'rp://api.minecraftservices.com/'
 const MinecraftServicesLogWithXbox = 'https://api.minecraftservices.com/authentication/login_with_xbox'
 const MinecraftServicesEntitlement = 'https://api.minecraftservices.com/entitlements/mcstore'
 const MinecraftServicesProfile = 'https://api.minecraftservices.com/minecraft/profile'
 
-const getFetchOptions = {
-  headers: {
-    'Content-Type': 'application/json',
-    'User-Agent': 'node-minecraft-protocol'
+// Initialize msal
+
+// See https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-common/docs/request.md#public-apis-1
+// for docs
+const msalConfig = {
+  auth: {
+    // the minecraft client:
+    // clientId: "000000004C12AE6F",
+    // clientId: '389b1b32-b5d5-43b2-bddc-84ce938d6737', // token from https://github.com/microsoft/Office365APIEditor
+    clientId: '60bad957-e36a-472d-a72a-612ad7f66b46', // @extremeheat
+    // clientId: '1fec8e78-bce4-4aaf-ab1b-5451cc387264', // MS Teams
+    authority: "https://login.microsoftonline.com/consumers"
+    // test out using live.com:
+    // validateAuthority: false,
+    // knownAuthorities: ["login.live.com"],
+    // protocolMode: 'OIDC',
+
+    // authority: 'https://login.live.com'
+  },
+  // cache: {
+  //   cachePlugin // your implementation of cache plugin
+  // },
+  system: {
+    loggerOptions: {
+      loggerCallback(loglevel, message, containsPii) {
+        // console.log('[msal] ', this.logLevel, message);
+      },
+      piiLoggingEnabled: false,
+      logLevel: msal.LogLevel.Verbose,
+    }
   }
 }
 
+let scopes = ['XboxLive.signin', 'offline_access']
+let msa = new MsaTokenManager({ msalConfig, scopes })
+let relayParty = 'rp://api.minecraftservices.com/'
```

Comment:
This variable is a duplicate of the const on L4, merge them?

Reply:
Yes, there needs to be a bit of refactoring to avoid code duplication. Still haven't decided how much code should be here and how much can be removed to a external package, like how the yggdrasil auth is currently done.